workspaces/spectacle/src/spectacle.ts

```diff
@@ -159,9 +159,12 @@ export async function makeSpectacle(opticContext: IOpticContext) {
         context: GraphQLContext
       ) => {
         const { diffId, captureId } = args;
-        await context
+        const {
+          onComplete,
+        } = await context
           .spectacleContext()
           .opticContext.capturesService.startDiff(diffId, captureId);
+        await onComplete;
```

Comment:
I don't think this should be necessary, and would stop us from receiving incremental diff results. As far as I understand, the diff queries all wait for the diff onComplete on the server side. So startDiff should always be able to return "immediately" and then queries could wait arbitrarily long. To experiment quickly with race conditions, you can add a stream transformer that adds delays to the fs write stream. @niclim

Reply:
The problem is, `onComplete` in the UI is immediately resolved after the startDiff spectacle call resolves https://github.com/opticdev/optic/blob/a4bdd09488ed0d071f1561ffa72212dd9bb61171/workspaces/spectacle-shared/src/local-cli/client.ts#L118-L122

This is causing our race condition with the fsSink. Because this resolves without `onComplete`, the trail learner (which reads from the FS) begins before the fsSink completes (even though the inMemory diff resolves wait for the inMemory passthrough to resolve).

Good point about the delay to the fswrite stream